File: exercise/leetcode/0207/0207.py
"""
两种思路：
1. 使用 BFS 
2. 使用 DFS
"""

import collections

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        """BFS + queue
        """
        graph = [[] for _ in range(numCourses)]
        indegree = [0 for _ in range(numCourses)]

        for item in prerequisites:
            graph[item[1]].append(item[0])
            indegree[item[0]] += 1
        
        s = collections.deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                s.append(i)

        cnt = 0
        while s:
            n = s.popleft()
            cnt += 1
            for v in graph[n]:
                indegree[v] -= 1
                if indegree[v] == 0:
                    s.append(v)
                
        return cnt == numCourses

    def canFinish2(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        """DFS + visit
        1. 为什么 visit 需要 -1，0，1 三种状态：
            暴力的 dfs 只需要两种状态 0 和 1，记录是否遍历过，但是为了节省时间，防止重复遍历，
            用第三种状态表示该节点在以前被遍历过，从该节点出发之后不可能有环，相当于已经被 check，
            就不用往下遍历了。
        """
        def dfs(cur) -> bool:
            if visit[cur] == 1:
                return False
            if visit[cur] == -1:
                return True
            
            visit[cur] = 1
            for v in graph[cur]:
                ret = dfs(v)
                if not ret:
                    return False
            visit[cur] = -1
            return True

        graph = [[] for _ in range(numCourses)]
        visit = [0 for _ in range(numCourses)]
        for item in prerequisites:
            graph[item[1]].append(item[0])

        for i in range(numCourses):
            ret = dfs(i)
            if not ret:
                return False
        return True

File: exercise/leetcode/0207/0207.cc
#include <vector>
#include <unordered_map>
#include <stack>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        unordered_map<int, vector<int>> road;
        for (int i = 0; i < prerequisites.size(); ++i) {
            road[prerequisites[i][1]].push_back(prerequisites[i][0]);
            indegree[prerequisites[i][0]] += 1;
        }
        
        vector<int> visited(numCourses, 0);
        stack<int> st;
        for (int i = 0; i < indegree.size(); ++i) {
            if (indegree[i] == 0) {
                st.push(i);
                visited[i] = 1;
            }
        }

        while (!st.empty()) {
            int id = st.top();
            st.pop();
            vector<int>& lst = road[id];
            for (int i=0; i<lst.size(); ++i) {
                indegree[lst[i]] -= 1;
                if (indegree[lst[i]] == 0) {
                    st.push(lst[i]);
                    visited[i] = 1;
                }
            }
        }

        for (int i : visited) {
            if (i == 0) {
                return false;
            }
        }
        return true;

    }
};